### PR TITLE
NOISSUE Curseforge makeover

### PR DESCRIPTION
**Pull request by phit**
makeover of the whole UI
add sorting
add version selection

image

#functional

### CODE REVIEW

**Comment by AverageComet250** [comment]
Why doesn't this exist already?

@peterix can you please merge this

**Comment by phit** [comment]
@AverageComet250 it's not done, the UI is kinda terrible, and the code is a mess

**Comment by AverageComet250** [comment]
Is there a way I can test this?

I'd love to try and get this pushed quicker, but I've never actually used Qt, so I can't help cleaning up the code, but I can either document it or test.

**Comment by phit** [comment]
once its ready to be tested it can go on the dev branch, then anyone can test it

**Comment by AverageComet250** [comment]
Nice

How long will that be?

**Comment by phit** [comment]
whenever it's ready

File: application/pages/modplatform/flame/FlameModel.cpp [hunk 1]
```diff
@@ -210,79 +210,17 @@ void Flame::ListModel::searchRequestFinished()
...
-            continue;
-        }
-        pack.broken = false;
+    QList<Flame::IndexedPack> newList;
+    auto packs = doc.array();
+    for(auto packRaw : packs) {
+        auto packObj = packRaw.toObject();
+
+        Flame::IndexedPack pack;
+        Flame::loadIndexedPack(pack, packObj);
```
Comment (jamierocks) [inline application/pages/modplatform/flame/FlameModel.cpp:219]: This call can throw an exception, so should be caught here :)

Here is what I did for modpacks.ch:
```cpp
    ModpacksCH::Modpack pack;
    try
    {
        ModpacksCH::loadModpack(pack, obj);
    }
    catch (const JSONValidationError &e)
    {
        qDebug() << QString::fromUtf8(response);
        qWarning() << "Error while reading pack manifest from FTB: " << e.cause();
        return;
    }
```

File: application/pages/modplatform/flame/FlamePage.cpp [hunk 2]
```diff
@@ -13,9 +13,19 @@ FlamePage::FlamePage(NewInstanceDialog* dialog, QWidget *parent)
     ui->setupUi(this);
     connect(ui->searchButton, &QPushButton::clicked, this, &FlamePage::triggerSearch);
     ui->searchEdit->installEventFilter(this);
-    model = new Flame::ListModel(this);
-    ui->packView->setModel(model);
+    listModel = new Flame::ListModel(this);
+    ui->packView->setModel(listModel);
+
+    ui->sortByBox->addItem(tr("Sort by featured"));
```
Comment (jamierocks) [inline application/pages/modplatform/flame/FlamePage.cpp:19]: Might be worth noting here that the index is a Curse implementation detail, and is important.

File: application/pages/modplatform/flame/FlamePage.cpp [hunk 3]
```diff
@@ -82,12 +94,48 @@ void FlamePage::onSelectionChanged(QModelIndex first, QModelIndex second)
...
+        {
+            QJsonParseError parse_error;
+            QJsonDocument doc = QJsonDocument::fromJson(*response, &parse_error);
+            if(parse_error.error != QJsonParseError::NoError) {
+                qWarning() << "Error while parsing JSON response from CurseForge at " << parse_error.offset << " reason: " << parse_error.errorString();
+                qWarning() << *response;
+                return;
+            }
+            QJsonArray arr = doc.array();
+            Flame::loadIndexedPackVersions(current, arr);
```
Comment (jamierocks) [inline application/pages/modplatform/flame/FlamePage.cpp:121]: Exceptions again :)

**Review by jamierocks — CHANGES_REQUESTED** [review]
Looks good :+1:

File: application/pages/modplatform/flame/FlamePage.cpp (hunk 2 above)
Comment (phit) [reply application/pages/modplatform/flame/FlamePage.cpp:21]: added a note

File: application/pages/modplatform/flame/FlamePage.cpp (hunk 3 above)
Comment (phit) [reply application/pages/modplatform/flame/FlamePage.cpp]: done

File: application/pages/modplatform/flame/FlameModel.cpp (hunk 1 above)
Comment (phit) [reply application/pages/modplatform/flame/FlameModel.cpp]: ah yup, some of the AtLauncher stuff I was copying is missing try/catch apparently

**Pull request merged**